### ISSUE TITLE
[repology] 'pname' is not exported to https://nixos.org/nixpkgs/packages-unstable.json.gz

### ISSUE THREAD

**Issue by volth** [NixOS/nixos-homepage#306]
**Describe the bug**
Although now majority of packages have 'pname' it is not exported to https://nixos.org/nixpkgs/packages-unstable.json.gz so repology is not able to pick `pname` and still parses `name` to `pname`+`version` using an unreliable heuristics.

https://github.com/repology/repology/issues/854#issuecomment-530094388

**To Reproduce**
`curl --silent https://nixos.org/nixpkgs/packages-unstable.json.gz`

**Comment by volth** [comment]
But it is even worse than current heuristics
https://github.com/NixOS/nix/blob/master/src/libexpr/names.cc#L19

It is not manually maintained `pname` but name.split('-').head emited
to json as though it is `pname`

**Comment by edolstra** [comment]
What is worse? The name/version splitting in https://github.com/NixOS/nix/blob/master/src/libexpr/names.cc is not a heuristic - it's the definition of what Nix considers the name and version parts.

**Comment by volth** [comment]
I mean we have to export to json manually defined `pname` and
`version` (if any defined).
Even if Nix do not use them.
Exporting result of splitting `name` has no sense, because they
provide no information and at the reader end could be calculated from
`name` (using copied from Nix, or possibly a smarter logic which would
produce package name better fit upstream namings than Nix's
consideration)
Whereas `pname` indicates how the package in called upstream in other
distros (and absence of `pname` indicates that the derivation is not a
package but somthing like buildEnv or fetchpatch)

**Comment by edolstra** [comment]
@AMDmi3 In Nix `python-3to2` is not a valid package name.

**Comment by volth** [comment]
> In Nix `python-3to2` is not a valid package name.

Well, but it is a valid upstream's (PYPI's, CPAN's, ..) `pname`.

Should those `pname`s (and `version`s) be escaped in Nix?

Something like.
```diff
- name = "${pname}-${version}";
+ name = "${escapePname pname}-${escapeVersion version};
```

where `escapePname` ensures there is no `-` following by a digit (for example by inserting `_` in between) and `escapeVersion` ensures the first char is a digit (by prefixing with `9999-` otherwise) 
?

**Comment by AMDmi3** [comment]
> @AMDmi3 In Nix `python-3to2` is not a valid package name.

Still it's used.

**Comment by AMDmi3** [comment]
I see now that the dump does contain `name` and `version` for all packages. This is great. A strange case was reported though (repology/repology-updater#941):

Though `trezor-udev-rules` [does](https://github.com/NixOS/nixpkgs/blob/4835c83e3a1a855015fdeb2b236b03d70c490110/pkgs/os-specific/linux/trezor-udev-rules/default.nix#L4-L5
) contain 

```
  pname = "trezor-udev-rules";
  version = "unstable-2019-07-17";
```

the values from the dump are different:

```
% curl --silent https://nixos.org/nixpkgs/packages-nixos-unstable.json.gz | gunzip | jq '.packages."trezor-udev-rules"' packages-unstable.json | egrep 'pname|version'
  "pname": "trezor-udev-rules-unstable",
  "version": "2019-07-17",
```

**Comment by volth** [comment]
Yes, the usefullness of how `pname` is exported to the dump has been
discussed: declared `pname` and `version` are not used, it is result
of `name` being splitted using a well-known function

**Comment by AMDmi3** [comment]
Sorry, two months have passed so I'm starting to forget the details. Then ping on that this needs to be fixed, as I'm receiving reports on "incorrect nix parsing".

**Comment by jtojnar** [comment]
There is an open [pull request](https://github.com/NixOS/nixpkgs/pull/68620) to actually expose the `pname` and `version` from the package expressions inside the `meta` attribute. We just cannot decide if we should fallback to the extracted values when the package only contains `name`, or just set them to `null`.

**Comment by volth** [comment]
@jtojnar It might have sense to let @ryantm make the design decisions. He got some EU funding for the update bot, so he is kind of full-time on the task. Matching Nix names to upstream name is a part of the bigger task of update automation.

**Comment by garbas** [comment]
as far as I can tell pname and version is now inlcuded in packages.json. please reopen if that is not the case.

**Comment by jtojnar** [comment]
Unfortunately, that are not the `pname` and `version` attributes passed to `mkDerivation` but rather the result of applying `parseDrvName` to the `name` attribute. That is pretty useless – Repology can implement `parseDrvName` (in fact it did) but getting the `pname` from the expression is not easily possible.

**Comment by FRidh** [comment]
Nix needs a name. At some point we introduced `pname` and `version` in Nixpkgs, that put together results in `name`. Nix splits the name part from the version part using `builtins.parseDrvName`. While one could argue that is unfortunate, the `pname` and `version` thing is Nixpkgs-specific. We can, and should, ensure `name` in derivations in Nixpkgs results in a correct name part and version part. We should test this happens by testing in `stdenv.mkDerivation` that, when `pname` is passed, `parseDrvName` returns an empty version.

**Comment by jtojnar** [comment]
I think it is fine when Nix uses combined `name` attribute for derivation name but trying to split it to get a version is an ugly hack. If some tools (¿just? `nix-env`) need version, they should recognize the `version` attribute. Then we can deprecate `builtins.parseDrvName`.

I do not think limiting what Nixpkgs can use in `pname` and `version` just so an ugly hack introduced for `nix-env` can work is a good idea.

**Comment by garbas** [comment]
I'm moving this ticket to https://github.com/NixOS/nixos-channel-scripts repository where this is being implemented. nixos-homepage repository and has a redirect in place.
